### PR TITLE
Implement dispose for wx simple TextEditor (base class)

### LINKED ISSUE #942: wx text style FontEditor needs a dispose method
After editing a text style `FontEditor` and closing the window (using demo `Standard_Editors/FontEditor_demo.py`), an error is thrown:
```
Traceback (most recent call last):
  File "/Users/icernyte/Documents/repos/traitsui/traitsui/wx/font_editor.py", line 291, in update_object
    self.value = self.control.GetValue()
AttributeError: 'NoneType' object has no attribute 'GetValue'
```

This happens because of a `FocusEvent` that is processed after `control` is set to `None`.

The event handler is bound in the general `TextEditor` that `FontEditor` subclasses, so the `dispose` method would actually have to be added there. I opened the issue for `FontEditor` because it's the only editor that throws an error like this. Not sure why others aren't running into the same issue.

### LINKED ISSUE #1145: UI disposal may cause AttributeError on None
I realize this issue keeps coming up in more context specific issues. I think it is worth a separate issue to
1) Collect these related issues (the symptoms)
2) Fix the root cause of the problem (the disease) rather than patching things.

The main symptom is that we get an error like this:

         "AttributeError: 'NoneType' object has no attribute ..." 

This could happen when the main UI window is being disposed, or when a nested UI is being disposed (say a list editor containing many nested UI is being updated). This symptom more frequently seen in tests, but had happened in production too.

Example of this symptom (as far as my memory can recall), note that there _might_ be misdiagnosis despite best efforts:
- https://github.com/enthought/traitsui/issues/854
- The first point in https://github.com/enthought/traitsui/issues/1051
- https://github.com/enthought/traitsui/issues/907
- The workaround in https://github.com/enthought/traitsui/pull/897
- The second traceback in https://github.com/enthought/traitsui/issues/1073

I also made some analysis earlier in Section "Problems with dispose" in this comment  https://github.com/enthought/traitsui/issues/431#issuecomment-648980689
(It is quite long, so I won't quote the whole thing here.)

Some digging into the history reveals that there was attempt to fix AttributeError like this, but the fix had to be reverted (Copied from https://github.com/enthought/traitsui/issues/854#issuecomment-644074983:)
>> Adding toolkit.destroy_control(self.control) to the dispose method of the editor.
>
> This was done in this PR a long time ago for avoiding the attribute error, by this PR 9 years ago #46
But then that is reversed in this PR: #283

The parts of code we want to be looking into would likely be `traitsui.ui.UI.dispose` and `traitsui.editor.Editor.dispose`.

This issue is toolkit agnostic, but I guess it will inevitably touch on toolkit specific code. In which case, these two issues would be relevant: https://github.com/enthought/traitsui/issues/431 (Qt), https://github.com/enthought/traitsui/issues/752 (Wx).

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #942 
Also resolve the second item in https://github.com/enthought/traitsui/issues/1073
Also part of #752 

The cause of the issue is as explained in #942.  This base class is used by both the text style FontEditor and the text style CompoundEditor.

I came across this while I was trying to _resolve_ #1145.
(The solution I am going to propose for #1145 also "resolves the error" in the context when the UI wrapping this editor has its `dispose` method called. But I believe removing the handler hooks with Unbind is still the right thing to do.)

As a drive-by improvement to the test coverage, implementations are added for testing the Qt implementation of the same editor.

### CODE REVIEW

File: traitsui/tests/editors/test_font_editor.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class TestFontEditor(unittest.TestCase):
+
+    def test_create_and_dispose_text_style(self):
+        # Setting focus on the widget and then disposing the widget
+        # should not cause errors.
+        view = View(Item("font_trait", style="text"))
+        tester = UITester([LOCAL_TARGET_REGISTRY])
+        with tester.create_ui(ObjectWithFont(), dict(view=view)) as ui:
+            wrapper = tester.find_by_name(ui, "font_trait")
+            wrapper.perform(command.MouseClick())
```
Comment (kitchoi) [inline traitsui/tests/editors/test_font_editor.py:53]: For the record, I was able to get the attribute error shown in #942, but for some reasons the event is processed so late that the error occurrs after the context manager exits. Hence it is not captured for causing the test to fail. The error was printed to the console however. I did not try to dive any further into understanding why the exception could not be captured. The fix resolves issue such that the error is not printed to the console.

File: traitsui/tests/editors/test_font_editor.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# A local registry of interaction handlers for testing purposes.
+LOCAL_TARGET_REGISTRY = TargetRegistry()
+
+if is_wx():
+    from traitsui.wx.font_editor import TextEditor
+    LOCAL_TARGET_REGISTRY.register_handler(
+        target_class=TextEditor,
+        interaction_class=command.MouseClick,
+        handler=lambda wrapper, _: wrapper.target.control.SetFocus()
+    )
```
Comment (kitchoi) [inline traitsui/tests/editors/test_font_editor.py:40]: This demonstrates how we can write tests for corner cases without necessarily exposing those test logic to the tester public API.
However this particular interaction might be worth bumping to the public API anyway.
After #1190, I will be able to extend support for the text-style font editor more trivially.

**Comment by kitchoi** [comment]
Updated to test the font editor for Qt as well. It is strictly speaking orthogonal to fixing #942, but is a good opportunity to expand test coverage and UITester support.

File: traitsui/tests/editors/test_font_editor.py (hunk 1 above)
Comment (aaronayres35) [reply traitsui/tests/editors/test_font_editor.py:39]: FWIW I observed the same

File: traitsui/wx/editor_factory.py [hunk 3]
```diff
@@ -116,6 +116,17 @@ def init(self, parent):
         parent.Bind(wx.EVT_TEXT_ENTER, self.update_object, id=self.control.GetId())
         self.set_tooltip()
 
+    def dispose(self):
```
Comment (aaronayres35) [inline traitsui/wx/editor_factory.py:119]: could we add a docstring here?

**Review by aaronayres35 — APPROVED** [review]
LGTM! just one minor comment for adding a docstring
